File: integrations/node/src/logger.ts
export type LogLevel = 'error' | 'info' | 'debug';

const levels: Record<LogLevel, number> = {
  error: 0,
  info: 1,
  debug: 2,
};

export interface Logger {
  error: (...args: unknown[]) => void;
  info: (...args: unknown[]) => void;
  debug: (...args: unknown[]) => void;
}

const validLevels = new Set<string>(Object.keys(levels));

export function isLogLevel(value: string): value is LogLevel {
  return validLevels.has(value);
}

export function createLogger(level: LogLevel): Logger {
  const threshold = levels[level];

  const noop = () => {};

  return {
    error: console.error.bind(console, '[dev-proxy]'),
    info: threshold >= levels.info ? console.log.bind(console, '[dev-proxy]') : noop,
    debug: threshold >= levels.debug ? console.log.bind(console, '[dev-proxy]') : noop,
  };
}

File: integrations/go/logger.go
package podproxy

import (
	"fmt"
	"os"
)

type logLevel int

const (
	levelError logLevel = iota
	levelInfo
	levelDebug
)

func parseLogLevel(value string) (logLevel, bool) {
	switch value {
	case "", "error":
		return levelError, true
	case "info":
		return levelInfo, true
	case "debug":
		return levelDebug, true
	}

	return levelError, false
}

type logger struct {
	level logLevel
}

func (l *logger) errorf(format string, args ...any) {
	l.writef(levelError, format, args...)
}

func (l *logger) infof(format string, args ...any) {
	l.writef(levelInfo, format, args...)
}

func (l *logger) debugf(format string, args ...any) {
	l.writef(levelDebug, format, args...)
}

func (l *logger) writef(level logLevel, format string, args ...any) {
	if level > l.level {
		return
	}

	fmt.Fprintf(os.Stderr, "[podproxy] "+format+"\n", args...)
}
